GAE/src/org/akvo/flow/rest/FormAssemblyRestService.java

```diff
@@ -0,0 +1,440 @@
+        question.setQuestionGroupId(question.getQuestionGroupId());
+        question.setSurveyId(questionDto.getSurveyId());
+        question.setVariableName(questionDto.getVariableName());
+        question.setOrder(questionDto.getOrder());
+        question.setMandatoryFlag(questionDto.getMandatoryFlag());
+        question.setPath(questionDto.getPath());
+        question.setAllowDecimal(questionDto.getAllowDecimal());
+        question.setAllowSign(questionDto.getAllowSign());
+        question.setMinVal(questionDto.getMinVal());
+        question.setMaxVal(question.getMaxVal());
+        question.setAllowExternalSources(questionDto.getAllowExternalSources());
+        question.setLocaleNameFlag(questionDto.getLocaleNameFlag());
+        question.setLocaleLocationFlag(questionDto.getLocaleLocationFlag());
+        question.setPersonalData(questionDto.getPersonalData());
+        question.setAllowPoints(questionDto.getAllowPoints());
+        question.setAllowLine(questionDto.getAllowLine());
+        question.setAllowPolygon(questionDto.getAllowPolygon());
+        question.setSourceQuestionId(questionDto.getSourceId());
+        question.setQuestionOptionMap(mapToOptions(questionDto));
+        return question;
+    }
+
+    @Nonnull
+    private TreeMap<Integer, QuestionOption> mapToOptions(QuestionDto questionDto) {
+        TreeMap<Integer, QuestionOption> mappedOptions = new TreeMap<>();
+        List<QuestionOptionDto> dtoList = questionDto.getOptionList();
+        if (dtoList != null) {
+            for (QuestionOptionDto questionOptionDto: dtoList) {
+                mappedOptions.put(questionOptionDto.getOrder(), mapToQuestionOption(questionOptionDto));
+            }
+        }
+        return mappedOptions;
+    }
+
+    private QuestionOption mapToQuestionOption(QuestionOptionDto questionOptionDto) {
+        QuestionOption questionOption = new QuestionOption();
+        questionOption.setKey(KeyFactory.createKey("QuestionOption", questionOptionDto.getKeyId()));
+        questionOption.setText(questionOptionDto.getText());
+        questionOption.setCode(questionOptionDto.getCode());
+        questionOption.setOrder(questionOptionDto.getOrder());
+        questionOption.setQuestionId(questionOptionDto.getQuestionId());
+        return questionOption;
+    }
+
+    private void attachCascadeResources(List<Question> questions) {
+        CascadeResourceDao cascadeResourceDao = new CascadeResourceDao();
+        for (Question question: questions) {
+            if (CASCADE.equals(question.getType())) {
+                CascadeResource cascadeResource = cascadeResourceDao.getByKey(question.getCascadeResourceId());
+                if (cascadeResource != null) {
+                    question.setCascadeResource(cascadeResource.getResourceId());
+                    question.setLevelNames(cascadeResource.getLevelNames());
+                }
+            }
+        }
+    }
+
+    private void attachFormTranslations(Survey form) {
```

Comment:
Since all the transations have a `parentId`, wouldnt it be simpler/more clear to create a map like `parentId -> (languageCode -> translationList)` and then process it from there, rather than having specific methods like `getFormTranslations()`, `getGroupTranslations()`?

Reply:
that's a really good idea, I have mapped them by parentId -> list of translations because the questions want list but all the others want a map